fix(consensus): Verify consensus branch ID in SIGHASH precomputation

upbqdn opened the pull request:
## Motivation

Close #9089.

### Specifications & References

#### [§ 4.10 SIGHASH Transaction Hashing](https://zips.z.cash/protocol/protocol.pdf#sighash)                                                                                
                                                                                           
> [NU5 only, pre-NU6] All transactions MUST use the NU5 consensus branch ID `0xF919A198` as defined in [ZIP-252].                                                                    
                                                                                           
> [NU6 only] All transactions MUST use the NU6 consensus branch ID `0xC8E71055` as defined in [ZIP-253].

[ZIP-252]: <https://zips.z.cash/zip-0252>
[ZIP-253]: <https://zips.z.cash/zip-0253>                                                  

## Solution

- Implement a check ensuring compliance with the consensus rules listed above. This check makes the tx SIGHASH precomputation panic if the tx contains an invalid value in its `nConsensusBranchId` field. This precomputation is implemented in `zebra-chain`, and is used by the tx verifier in `zebra-consensus` and by the transparent script verifier in `zebra-script`, but this verifier is only called from the tx verifier. It might be worth optimizing this, but that is currently not possible, as described in #9165. The panic should never occur in regular use because the tx verifier implements another consensus branch ID check according to another consensus rule implemented here: https://github.com/zcashfoundation/zebra/blob/05a9b6171ccbc06bcc888470e4be8a60b1d574ef/zebra-consensus/src/transaction/check.rs#L513-L552, which takes effect early on and doesn't panic.
- Add new error types in `zebra-chain`.
- Pass `NetworkUpgrade` instead of `ConsensusBranchId` to the SIGHASH precomputation and tx conversion.
- Use `TryFrom` instead of ad-hoc functions for conversion.
- Remove a redundant `TryFrom<&Transaction> for zcash_primitives::transaction::Transaction`.
- Document that we now implement the consensus rules listed above.
- Impl `NetworkUpgrade::previous_upgrade` and refactor `NetworkUpgrade::next_upgrade` so that it doesn't explicitly enumerate all upgrades.
- Don't make `NETWORK_UPGRADES_IN_ORDER` `pub`.

### Tests

- Add tests that check that tx SIGHASH precomputation and computation fail when a tx attempts to use an invalid consensus branch id. 

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

upbqdn committed:
Add `has_foo` fns to `Transaction`

upbqdn committed:
Add V5 SIGHASH test based on consensus branch ID

upbqdn committed:
Merge branch 'main' into fix-sighash

upbqdn commented on zebrad/src/components/mempool/downloads.rs:
```diff
@@ -373,6 +373,7 @@ where
                 .oneshot(tx::Request::Mempool {
                     transaction: tx.clone(),
                     height: next_height,
+                    skip_checks: None
```
Why does guarding this by `#[cfg(any(test, feature = "proptest-impl"))]` not work, and why does it work as it is since `skip_checks` is guarded by the same attribute in `zebra-consensus`?

upbqdn replied:
Turns out it doesn't. I have to understand better how feature assignment behaves during compilation. And it's probably not a good idea to do it this way anyway.

upbqdn replied:
I used this hack to solve this: https://github.com/zcashfoundation/zebra/blob/9597792567982ae19643eba0d2537d3a81fe5956/zebrad/Cargo.toml#L301-L304.

upbqdn replied:
I'm trying to think of something better in general.

upbqdn committed:
Guard `skip_checks` by test features

upbqdn committed:
Enable `proptest-impl` for `zebrad` in tests

upbqdn committed:
Simplify conditional compilation

upbqdn committed:
Enable `proptest-impl` in scanner's dev deps

upbqdn committed:
Fix conditional compilation in `zebra-chain` tests

oxarbitrage commented on zebra-consensus/src/transaction.rs:
```diff
@@ -575,13 +600,13 @@ where
+                        miner_fee.expect("fee should have been checked earlier"),
                         legacy_sigop_count,
+                        #[cfg(any(test, feature = "proptest-impl"))]
+                        req.skip_checks().is_some_and(|checks| checks.contains(&SkipCheck::Zip317))
```
I am not sure why but i remember seeing other odd behaviors when using `proptest-impl`. Here is one option to fix it:


```suggestion
                        #[cfg(any(test, feature = "proptest-impl"))]
                        req.skip_checks().is_some_and(|checks| checks.contains(&SkipCheck::Zip317)),
                        #[cfg(not(any(test, feature = "proptest-impl")))]
                        false,
```

upbqdn replied:
Nope.

oxarbitrage replied:
Yea, sorry about that. I will try something else.

upbqdn committed:
Add error types for `zebra-chain`

upbqdn committed:
`impl TryFrom<u32> for NetworkUpgrade`

upbqdn committed:
`impl TryFrom<ConsensusBranchId> for BranchId`

upbqdn committed:
Rm `fn from_branch_id() -> Option<NetworkUpgrade>`

upbqdn committed:
Check consensus branch ID in SIGHASH computation

upbqdn committed:
Simplify tx deserialization

upbqdn committed:
Rm `impl TryFrom<&Trans> for zp_tx::Trans`

upbqdn committed:
Update tests

upbqdn committed:
Update tests

upbqdn committed:
Add docs for `to_librustzcash`

upbqdn committed:
Update docs for `PrecomputedTxData::new`

upbqdn committed:
Document the SIGHASH consensus rules we missed

upbqdn committed:
Update docs for script validation

upbqdn committed:
Merge branch 'main' into fix-sighash

upbqdn committed:
Fix script verification tests

In a previous commit, I erroneously edited the tests so that they'd
expect `Ok`s instead of `Err`s. This commit fixes that.

upbqdn committed:
Fix spelling

upbqdn committed:
Impl `NetworkUpgrade::iter()`

upbqdn committed:
Refactor `Network_upgrade::next_upgrade`

upbqdn committed:
Impl `NetworkUpgrade::previous_upgrade`

upbqdn committed:
Impl `Transaction::hash_shielded_data`

upbqdn committed:
Don't make  `NETWORK_UPGRADES_IN_ORDER` `pub`

upbqdn committed:
Test `Transaction::sighash` with cons branch ids

upbqdn committed:
Merge branch 'main' into fix-sighash

upbqdn committed:
Extend the `consensus_branch_id` test

upbqdn committed:
Derive `Debug` for `SigHasher`

upbqdn committed:
Remove the beautiful test for tx verifier

upbqdn committed:
Remove the "skip check" functionality

upbqdn committed:
Revert the compilation adjustments

arya2 commented on zebra-chain/src/parameters/network_upgrade.rs:
```diff
@@ -350,17 +350,18 @@ impl NetworkUpgrade {
+        Self::iter()
+            .position(|nu| self == nu)
+            .filter(|&p| p > 0)
+            .map(|p| NETWORK_UPGRADES_IN_ORDER[p - 1])
```
Optional/Nitpick, requires a suggestion below:

```suggestion
        Self::iter().skip_while(|&nu| self != nu).nth(1)
    }

    /// Returns the previous network before after this network upgrade
    pub fn previous_upgrade(self) -> Option<Self> {
        Self::iter().rev().skip_while(|&nu| self != nu).nth(1)
```

upbqdn replied:
Amazing. Thanks.

arya2 commented on zebra-chain/src/parameters/network_upgrade.rs:
```diff
@@ -538,6 +539,11 @@ impl NetworkUpgrade {
     }
+
+    /// Returns an iterator over [`NetworkUpgrade`] variants.
+    pub fn iter() -> impl Iterator<Item = NetworkUpgrade> {
```
Required for the suggestion above
```suggestion
    pub fn iter() -> impl DoubleEndedIterator<Item = NetworkUpgrade> {
```

upbqdn committed:
Apply suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

upbqdn committed:
Fix docs

arya2 commented on zebra-chain/src/transaction/tests/vectors.rs:
```diff
@@ -801,6 +803,46 @@ fn zip244_sighash() -> Result<()> {
+                .expect_err("creating the sighasher panics under nu other than the tx one");
+
+            std::panic::catch_unwind(|| tx.sighash(any_other_nu, HashType::ALL, &[], None))
+                .expect_err("the sighash computation panics under nu other than the tx one");
```
Could you mention this in the method docs for `sighash()` and `sighasher()` too?

upbqdn replied:
This is a bit more nuanced. It won't panic for txs that don't contain the `nConsensusBranchId` field. This test uses only those, so the message here can be simplified. Other reasons for panicking can be issues with obtaining a librustzcash version of the consensus branch id from the passed network upgrade. 

This is documented in the corresponding functions: https://github.com/zcashfoundation/zebra/blob/2e32fd688c53e6956d0a7447bf750f37a7df0a99/zebra-chain/src/transaction.rs#L1237 and https://github.com/zcashfoundation/zebra/blob/2e32fd688c53e6956d0a7447bf750f37a7df0a99/zebra-chain/src/primitives/zcash_primitives.rs#L243.
We usually document panics only in the functions from which they originate, so I thought it was fine as it was. 

I can document it in those functions you mentioned, but then I would have to describe the whole story, which feels too repetitive. Would you prefer that?

arya2 replied:
> We usually document panics only in the functions from which they originate, so I thought it was fine as it was.

I think it'd be nice to say "Panics on parser errors or network upgrade mismatches" with a reference to the docs for `PrecomputedTxData::new()` + `to_librustzcash()`, but I also think it's fine as-is and definitely shouldn't be repeated.

upbqdn replied:
Addressed in 3b25023db93e708837da4e6d625810e932e81869.

upbqdn committed:
Clarify panic conditions in docs

arya2 commented on zebra-consensus/src/transaction.rs:
```diff
@@ -575,12 +575,10 @@ where
                     let transaction = VerifiedUnminedTx::new(
-                        transaction,
-                        miner_fee.expect(
-                            "unexpected mempool coinbase transaction: should have already rejected",
-                        ),
+                        tx.clone(),
```
Why this change?

upbqdn replied:
Thanks for spotting this. It was required for the conditional compilation I had earlier, which is gone now. I'll revert it.

arya2 commented on zebra-script/src/lib.rs:
```diff
@@ -451,13 +440,13 @@ mod tests {
+            .expect_err("verification should fail");
+        verifier
+            .is_valid(NetworkUpgrade::Blossom, input_index + 1)
+            .expect_err("verification should fail");
```
```suggestion
        verifier
            .is_valid(NetworkUpgrade::Blossom, input_index + 1)
            .expect_err("verification should fail");
```

upbqdn replied:
This shouldn't have been merged. The verifier is supposed to fail twice. It's in the name of the test, similarly to other tests in the file.

upbqdn replied:
I'll fix it in another PR.

oxarbitrage replied:
sorry about that, i thought it was an oversight.

arya2 reviewed (dismissed):
Looks good, thank you!

oxarbitrage committed:
remove duplicated verification

Co-authored-by: Arya <aryasolhi@gmail.com>
